chain/xrp/client/client.go

```diff
@@ -0,0 +1,718 @@
+	Account            string          `json:"Account"`
+	Amount             string          `json:"Amount,omitempty"`
+	Destination        string          `json:"Destination,omitempty"`
+	Fee                string          `json:"Fee"`
+	Flags              int64           `json:"Flags"`
+	LastLedgerSequence int64           `json:"LastLedgerSequence"`
+	Sequence           int64           `json:"Sequence"`
+	SigningPubKey      string          `json:"SigningPubKey"`
+	TransactionType    string          `json:"TransactionType"`
+	TxnSignature       string          `json:"TxnSignature"`
+	Hash               string          `json:"hash"`
+	DeliverMax         string          `json:"DeliverMax,omitempty"`
+	TakerGets          TakerGets       `json:"TakerGets,omitempty"`
+	TakerPays          TakerPays       `json:"TakerPays,omitempty"`
+	CtID               string          `json:"ctid,omitempty"`
+	Meta               TransactionMeta `json:"meta"`
+	Validated          bool            `json:"validated"`
+	Date               int64           `json:"date"`
+	LedgerIndex        int64           `json:"ledger_index"`
+	InLedger           int64           `json:"inLedger"`
+	Status             string          `json:"status"`
+}
+
+type TakerGets struct {
+	StringValue string  `json:"-"`
+	AmountValue *Amount `json:"-"`
+	IsString    bool    `json:"-"`
+}
+
+// UnmarshalJSON is the custom unmarshal method for TakerGets
+func (tg *TakerGets) UnmarshalJSON(data []byte) error {
+	var str string
+	if err := json.Unmarshal(data, &str); err == nil {
+		tg.StringValue = str
+		tg.IsString = true
+		return nil
+	}
+
+	var amount Amount
+	if err := json.Unmarshal(data, &amount); err == nil {
+		tg.AmountValue = &amount
+		tg.IsString = false
+		return nil
+	}
+
+	return fmt.Errorf("TakerGets is neither a string nor an Amount")
+}
+
+type TakerPays struct {
+	StringValue string  `json:"-"`
+	AmountValue *Amount `json:"-"`
+	IsString    bool    `json:"-"`
+}
+
+// UnmarshalJSON is the custom unmarshal method for TakerPays
+func (tp *TakerPays) UnmarshalJSON(data []byte) error {
+	var str string
+	if err := json.Unmarshal(data, &str); err == nil {
+		tp.StringValue = str
+		tp.IsString = true
+		return nil
+	}
+
+	var amount Amount
+	if err := json.Unmarshal(data, &amount); err == nil {
+		tp.AmountValue = &amount
+		tp.IsString = false
+		return nil
+	}
+
+	return fmt.Errorf("TakerGets is neither a string nor an Amount")
+}
+
+type TransactionMeta struct {
+	AffectedNodes     []AffectedNodes `json:"AffectedNodes"`
+	TransactionIndex  int64           `json:"TransactionIndex"`
+	TransactionResult string          `json:"TransactionResult"`
+	DeliveredAmount   string          `json:"delivered_amount,omitempty"`
+}
+
+type AffectedNodes struct {
+	CreatedNode  *CreatedNode  `json:"CreatedNode,omitempty"`
+	ModifiedNode *ModifiedNode `json:"ModifiedNode,omitempty"`
+}
+
+// UnmarshalJSON for AffectedNode
+func (an *AffectedNodes) UnmarshalJSON(data []byte) error {
```

Comment:
why the custom unmarshal?  if the default unmarshal doesn't work, would like to see comment to understand what's causing issues.

Reply:
I'd say it is necessary as all those custom UnmarshalJSON functions I've created are targeting the case when the field on which they are implemented can have different data types depending on the value.
.
The default Go UnmarshalJSON won't work ( I've also tested it ) as it cannot handle multiple types for a single field automatically. Also it expects a field to have a specific type and if it encounters a type mismatch it will throw error.
.
A viable example is with AffectedNodes in here which can hold elements of both CreatedNode and ModifiedNode.

Reply:
To me it looks like it implements the behavior that is default already.   I.e. `"CreatedNode"`  always maps to decode as `CreatedNode`, and `"ModifiedNode"` always maps to decode as `ModifiedNode`.